### PR TITLE
[HINTS] Find element hint

### PR DESCRIPTION
**Pull request by Juan-M-V**
# Find element hint implementation

## Description

Implements hint needed for find_element https://github.com/starkware-libs/cairo-lang/blob/2abd303e1808612b724bc1412b2b5babd04bb4e7/src/starkware/cairo/common/find_element.cairo#L29-L59

## Checklist
- [x] Unit tests added
- [x] Integration tests added.

### CODE REVIEW

File: src/vm/hints/find_element_hint.rs [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+            Ok(Some(maybe_rel_n_elms)),
+            Ok(_),
+            Ok(Some(maybe_rel_key)),
+        ) => {
+            for (name, builtin) in &vm.builtin_runners {
+                //Check that range_check_builtin is present
+                let _builtin = match builtin.as_any().downcast_ref::<RangeCheckBuiltinRunner>() {
+                    Some(b) => b,
+                    None => return Err(VirtualMachineError::NoRangeCheckBuiltin),
+                };
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:114]: This fails whenever `RangeCheckBuiltinRunner` is not the first runner in the vector, which is most likely not what we intend.

File: src/vm/hints/find_element_hint.rs [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+
+                if name == &"range_check".to_string() {
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:116]: This part can become (will need to update everything below):
```suggestion
            let builtin = vm.builtin_runners.iter()
                .find(|(name, _)| name.as_str() == "range_check")
                .ok_or(VirtualMachineError::NoRangeCheckBuiltin)?
                .1
                .as_any().downcast_ref::<RangeCheckBuiltinRunner>()
                .map_err(VirtualMachineError::NoRangeCheckBuiltin)?;
```

File: src/vm/hints/find_element_hint.rs [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+                        return Err(VirtualMachineError::ExpectedInteger(
+                            maybe_rel_elm_size.clone(),
+                        ));
+                    };
+
+                    if !elm_size.is_positive() {
+                        return Err(VirtualMachineError::ValueOutOfRange(elm_size.clone()));
+                    }
+
+                    if let Some(find_element_index_value) = vm.find_element_index.clone() {
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:129]: Why not a reference?

File: src/vm/errors/vm_errors.rs [hunk 4]
```diff
@@ -61,6 +61,10 @@ pub enum VirtualMachineError {
     NoDictTracker(usize),
     NoValueForKey(BigInt),
     AssertLtFelt(BigInt, BigInt),
+    FindElemKeyNotFound(MaybeRelocatable),
+    FindElemMaxSize(BigInt, BigInt),
+    InvalidIndex(BigInt, MaybeRelocatable, MaybeRelocatable),
+    FindElemNoFoundKey,
```
Comment (Oppen) [inline src/vm/errors/vm_errors.rs:67]: `KeyNotFound`

File: src/vm/errors/vm_errors.rs [hunk 5]
```diff
@@ -61,6 +61,10 @@ pub enum VirtualMachineError {
     NoDictTracker(usize),
     NoValueForKey(BigInt),
     AssertLtFelt(BigInt, BigInt),
+    FindElemKeyNotFound(MaybeRelocatable),
+    FindElemMaxSize(BigInt, BigInt),
```
Comment (Oppen) [inline src/vm/errors/vm_errors.rs:65]: Reuse errors here.

File: src/vm/hints/find_element_hint.rs (hunk 3 above)
Comment (Oppen) [reply src/vm/hints/find_element_hint.rs:129]: `find_element_index` should probably be part of an exec scope.

File: src/vm/hints/find_element_hint.rs [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+
+                        let array_start = vm
+                            .memory
+                            .get(&array_ptr_addr)
+                            .map_err(VirtualMachineError::MemoryError)?
+                            .ok_or(VirtualMachineError::FindElemNoFoundKey)?;
+
+                        for i in 0..n_elms_iter {
+                            let iter_addr =
+                                &array_start.add_int_mod(&(elm_size * bigint!(i)), &vm.prime)?;
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:195]:
```suggestion
                        let mut array_start = vm
                            .memory
                            .get(&array_ptr_addr)
                            .map_err(VirtualMachineError::MemoryError)?
                            .ok_or(VirtualMachineError::FindElemNoFoundKey)?;

                        for i in 0..n_elms_iter {
                            array_start = array_start.add_int_mod(elm_size, &vm.prime)?;
```

File: src/vm/hints/find_element_hint.rs [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+        );
+
+        assert_eq!(
+            vm.memory.get(&MaybeRelocatable::from((0, 3))),
+            Ok(Some(&MaybeRelocatable::Int(bigint!(1))))
+        )
+    }
+
+    #[test]
+    fn element_found_by_global() {
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:383]:
```suggestion
    fn element_found_by_oracle() {
```

File: src/vm/hints/find_element_hint.rs [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+        assert_eq!(
+            vm.memory.get(&MaybeRelocatable::from((0, 3))),
+            Ok(Some(&MaybeRelocatable::Int(bigint!(1))))
+        )
+    }
+
+    #[test]
+    fn element_not_found_search() {
+        let (mut vm, ids) = init_vm_ids();
+        vm.memory.data[0][4] = Some(MaybeRelocatable::from(bigint!(7)));
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:401]: Try not to access `memory.data` directly. We want to abstract that.

File: src/vm/hints/find_element_hint.rs [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+        vm.builtin_runners.push((
+            "output".to_string(),
+            Box::new(OutputBuiltinRunner::new(true)),
+        ));
+
+        assert_eq!(
+            execute_hint(&mut vm, FIND_ELEMENT_HINT, ids, &ApTracking::new()),
+            Err(VirtualMachineError::NoRangeCheckBuiltin)
+        );
+    }
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:488]: Add test with range check as second builtin. Should trigger the bug mentioned above.

File: src/vm/vm_core.rs [hunk 10]
```diff
@@ -65,6 +65,8 @@ pub struct VirtualMachine {
     current_step: usize,
     skip_instruction_execution: bool,
     pub dict_manager: DictManager,
+    pub find_element_index: Option<BigInt>,
```
Comment (Oppen) [inline src/vm/vm_core.rs:68]: Put into exec_scope.

File: src/vm/hints/find_element_hint.rs (hunk 1 above)
Comment (Juan-M-V) [reply src/vm/hints/find_element_hint.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/25fad2bc0eb02e2dd81e354dea1e506c6c2373b5

File: src/vm/hints/find_element_hint.rs (hunk 2 above)
Comment (Juan-M-V) [reply src/vm/hints/find_element_hint.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/25fad2bc0eb02e2dd81e354dea1e506c6c2373b5

File: src/vm/errors/vm_errors.rs (hunk 4 above)
Comment (Juan-M-V) [reply src/vm/errors/vm_errors.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/96eecc23a491a0e48bd196b069ee3203b383a564

File: src/vm/errors/vm_errors.rs (hunk 5 above)
Comment (Juan-M-V) [reply src/vm/errors/vm_errors.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/96eecc23a491a0e48bd196b069ee3203b383a564

File: src/vm/hints/find_element_hint.rs (hunk 7 above)
Comment (Juan-M-V) [reply src/vm/hints/find_element_hint.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/365075dac6f3ea4030c0d87b5e33071a326d51fa

File: src/vm/hints/find_element_hint.rs (hunk 3 above)
Comment (Juan-M-V) [reply src/vm/hints/find_element_hint.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/24979a3c5067e6b1097e6e7555da60aee8639164

File: src/vm/hints/find_element_hint.rs (hunk 8 above)
Comment (Juan-M-V) [reply src/vm/hints/find_element_hint.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/44039cc5f34fe872d57d50c8bf268eed77b0fc4c

File: src/vm/hints/find_element_hint.rs (hunk 6 above)
Comment (Juan-M-V) [reply src/vm/hints/find_element_hint.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/294/commits/e8eb2b10e7b4857dc441248cde0d6b0c3944db21

File: src/vm/hints/find_element_hint.rs [hunk 11]
```diff
@@ -106,115 +34,132 @@ pub fn find_element(
...
+                }
+
+                if let Some(variables) = vm.exec_scopes.get_local_variables() {
+                    if let Some(PyValueType::BigInt(find_element_max_size)) =
+                        variables.get("find_element_max_size")
+                    {
+                        if n_elms > find_element_max_size {
+                            return Err(VirtualMachineError::FindElemMaxSize(
+                                find_element_max_size.clone(),
+                                n_elms.clone(),
```
Comment (Oppen) [inline src/vm/hints/find_element_hint.rs:122]: Shouldn't this be tested unconditionally at the beginning of the hint?

Comment (Juan-M-V) [reply src/vm/hints/find_element_hint.rs:122]: Apparently not https://github.com/starkware-libs/cairo-lang/blob/2abd303e1808612b724bc1412b2b5babd04bb4e7/src/starkware/cairo/common/find_element.cairo#L48-L51
It's tested only if there is an element_max_size

Comment (Oppen) [reply src/vm/hints/find_element_hint.rs:122]: Sorry, I meant unconditionally regarding whether or not an element index was passed.

**Review by Oppen — APPROVED** [review]

**Pull request merged**